clang/lib/Driver/ToolChains/SYCL.cpp

```diff
@@ -270,6 +270,10 @@ const char *SYCL::Linker::constructLLVMLinkCommand(
     // TODO: temporary workaround for a problem with warnings reported by
     // llvm-link when driver links LLVM modules with empty modules
     CmdArgs.push_back("--suppress-warnings");
+    // FIXME remove this when opaque pointers are supported for SPIR-V
+    if (!this->getToolChain().getTriple().isSPIR()) {
+      CmdArgs.push_back("-opaque-pointers");
+    }
```

Comment:
`llvm-link` takes the `opaque-pointers` preference of the first bc file passed to it. So this will fail:

```
llvm-link /usr/local/cuda/nvvm/libdevice/libdevice.10.bc /path/to/libclc_ptx.bc
```
Since llvm-link detects that opaque pointers are not going to be used (since `libdevice.10.bc` uses typed pointers) and then it links with `libclc_ptx.bc` which uses opaque pointers. Linking opaque ptrs bc in typed pointer mode fails because llvm-link cannot change from opaque to typed ptrs.

Using the same invocation but swapping the order of the bc files will work, since `llvm-link` then uses opaque pointer mode, which can happily link with typed pointer bc, since it is easy to convert a typed ptr bc into opaque pointer bc.

We need to add the `opaque-pointers` flag at least in the SYCL libdevice linking, since in this case, we have typed bc (SYCL libdevice modules like `libsycl-fallback-cassert-e2fdb3.cubin`) being linked with libclc (opaque) and CUDA libdevice modules.

My reason for making this a global option for llvm link for non SPIR-V targets, is that there may be corner cases where the user might want to link some bc that is typed, and the typed bc comes before the opaque ptr modules that are generated by normal compilation. I have not been able to generate a pathological case using for instance
```
clang++ -fsycl test.cpp -fsycl-targets=nvptx64-nvidia-cuda -Xclang -mlink-builtin-bitcode -Xclang typed_ptr.bc
```
Since the linked bc currently comes at the end of the `llvm-link` invocation. This is probably reliable behaviour so maybe we don't need to worry about this.

Anyway I have changed the behaviour so that the `-opaque-pointers` flag is only added when linking SYCL libdevice (which is only supported for PTX and SPIRV targets, so the change will not affect AMDGCN).

Let me know your thoughts